IP Whitelists for Frontend (with Docker- & Kubernetes-Provider Support)

MaZderMind opened the pull request:
This PR fixes #816. It adds the ability to specify a List of CIDR Whitelists per Frontend. If Whitelists are configured and the Request-IP does not match one of them, it is aborted with a 403 Forbidden.

This obviously requires traefik to be directly mapped on public ips, if the incoming requests are nat'ed, ie through dockers publish-functionality or docker-compose ports, traefik will only see an ip of the internal network and won't be able to filter according to the source-ip.

This PR adds support for the Docker-Provider and the Kubernetes-Provider, but it should be trivial to extend the support to other Providers, I just have no test-setup for those. For both it parses the Annoatation/Label "traefik.frontend.whitelistSourceRange", for Kubernetes it also accepts the quasi-standard "ingress.kubernetes.io/whitelist-source-range". They can take a Comma-Separated List of CIDR Nets (IPv4 & v6), like this:
```yml
machine:
  image: katacoda/docker-http-server
  labels:
    - "traefik.backend=local-only-echo"
    - "traefik.frontend.whitelistSourceRange="10.42.0.0/16, 152.89.1.33/32, afed:be44::/16"
    - "traefik.frontend.rule=Host:local-only-echo.example.com"
```

or for Kubernetes:
```yml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: hello-minikube
  annotations:
    traefik.frontend.whitelistSourceRange: "10.42.0.0/16, 152.89.1.33/32, afed:be44::/16"
spec:
  rules:
  - host: echo.kube
    http:
      paths:
      - path: /
        backend:
          serviceName: hello-minikube
          servicePort: web
```

For File-Based Config, there's the `whitelistSourceRange` property, which takes an array of CIDR Net-Strings:
```toml
[frontends]
  [frontends.frontend2]
  backend = "backend1"
  passHostHeader = true
  priority = 10
  whitelistSourceRange = ["10.42.0.0/16", "152.89.1.33/32", "afed:be44::/16"]
```

The applied Whitelists are Displayed on the Dashboard:
![whitelists-dashboard](https://cloud.githubusercontent.com/assets/142237/24220320/1c183ffe-0f4a-11e7-9407-94a139c704b9.png)

For my own personal use and for easy testing, I pushed a Docker-Image based on this PR:
https://hub.docker.com/r/mazdermind/traefik/tags/

I'd be happy to receive some feedback, because Go is still fairly new.
Best Regards,
Peter

timoreimann commented:
@MaZderMind thanks for your contribution!

I'll try to review the code. In the meantime, could you please run `make fmt` and commit your changes in order to please the style gods (and our CI)?

MaZderMind commented:
@timoreimann Thank you, I'm already at it. Sorry for not checking this before pushing, I'm quite new to the Go Ecosystem

timoreimann commented:
@MaZderMind no worries. The CI is our friend. :-)

timoreimann commented on provider/docker_test.go:
```diff
@@ -956,9 +999,10 @@ func TestDockerLoadDockerConfig(t *testing.T) {
+					Backend:              "backend-test",
+					PassHostHeader:       true,
+					EntryPoints:          []string{},
+					WhitelistSourceRange: []string{},
```
Do we really need to add an empty white lists field here and in the many other tests? On first look, it seems like the tests should work fine without them.

MaZderMind replied:
Hmm Unfortunately reflect.DeepEqual distinguished from a []string{} initialized with nil and one initialized as a empty list, which took me quite a while because the tests fail with the unhelpful message:
```
expected map[string]*types.Frontend{"frontend-Host-test1-docker-localhost":(*types.Frontend)(0xc4201c49c0), "frontend-Host-test2-docker-localhost":(*types.Frontend)(0xc4201c4a20)},
     got map[string]*types.Frontend{"frontend-Host-test1-docker-localhost":(*types.Frontend)(0xc4202e8480), "frontend-Host-test2-docker-localhost":(*types.Frontend)(0xc4202e8600)}
```

I'd suppose it would be best to remove the reflect.DeepEqual shortcut and instead assert the expectations one by one, which would also lead to clearer messages in case of a failing test, but I don't think this should be part of this PR.

Given the above preconditions, it is indeed necessary to initialize WhitelistSourceRange with an empty style everywhere.

timoreimann replied:
I'd still remove the empty slice from the test case definition and sanitize inside the test body; that is, check if the value is nil and assign the empty slice to it. That keeps the test cases clean and moves the sanitization logic into a central location.

As far as being able to easily spot what's wrong in case of a test failure, you should use something like [go-spew](https://github.com/davecgh/go-spew)'s `Sdump`. We already use it at other occasions, so you should be able to use it in your tests too.

timoreimann commented on provider/kubernetes.go:
```diff
@@ -27,6 +27,9 @@ const (
 	ruleTypePathPrefix         = "PathPrefix"
+
+	annotationKubernetesWhitelistSourceRange = "ingress.kubernetes.io/whitelist-source-range"
+	annotationTraeffikWhitelistSourceRange   = "traefik.frontend.whitelistSourceRange"
```
Do we actually have to support this alternative format? Can't we just support the more Kube-idiomatic annotation above?

timoreimann commented on provider/kubernetes.go:
```diff
@@ -160,12 +163,27 @@ func (provider *Kubernetes) loadIngresses(k8sClient k8s.Client) (*types.Configur
+					if len(s) > 0 {
+						whitelistSourceRange = append(whitelistSourceRange, s)
+					}
+				}
```
We have this function elsewhere already. Can we outsource it into a central location? Maybe a utility package if we don't have one yet?

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1609,6 +1638,19 @@ func TestIngressAnnotations(t *testing.T) {
 					Method: "wrr",
 				},
 			},
+			"test/ip-source-range": {
```
Let's be consistent in the naming and say `whitelist-source-range` here.

timoreimann commented on provider/kubernetes_test.go:
```diff
@@ -1545,6 +1545,35 @@ func TestIngressAnnotations(t *testing.T) {
+								Paths: []v1beta1.HTTPIngressPath{
+									{
+										Path: "/ip-source-range",
+										Backend: v1beta1.IngressBackend{
```
s/ip/whitelist/ for naming consistency.

timoreimann commented on server.go:
```diff
@@ -740,6 +740,17 @@ func (server *Server) loadConfig(configurations configs, globalConfiguration Glo
 						}
+
+						if len(frontend.WhitelistSourceRange) > 0 {
+							log.Info("Configured IP Whitelists: ", frontend.WhitelistSourceRange)
```
I'd place this log messages only after we are clear there won't be an error returned by `NewIPWhitelister`.

timoreimann replied:
Please use `log.Infof` and apply the correct verb (`%s` I'd think).

timoreimann commented on server.go:
```diff
@@ -740,6 +740,17 @@ func (server *Server) loadConfig(configurations configs, globalConfiguration Glo
+							ipSourceRanges := frontend.WhitelistSourceRange
+							ipWhitelistMiddleware, err := middlewares.NewIPWhitelister(ipSourceRanges)
+							if err != nil {
+								log.Fatal("Error creating IP Whitelister: ", err)
```
Please use `log.Fatalf`.

timoreimann commented on server.go:
```diff
@@ -740,6 +740,17 @@ func (server *Server) loadConfig(configurations configs, globalConfiguration Glo
+							}
+							negroni.Use(ipWhitelistMiddleware)
+						}
+
```
This part needs to be covered by a test. I know it's tricky because we have zero tests yet, but we need to start somewhere.

For a starter, [here are my tests from a different PR](https://github.com/containous/traefik/blob/539e22a37fbe28f4f7f93e5825960fee7a714990/server_test.go).

MaZderMind replied:
The Server-Construction-Code is not really nice to be tested, is it? ;)
I extracted the configuration of the middleware into a smaller, easy to test function and tested that. I think this is the better approach instead of configuring a complete Server and trying to fiddle out which middlewares are configured.

timoreimann commented on middlewares/ip_witelister.go:
```diff
@@ -0,0 +1,84 @@
+			log.Debugf("Source-IP %s matched none of the %d whitelists, rejecting", remoteIP, len(whitelists))
+			reject(w)
+		}
+	})
```
The middleware needs testing. Once you have moved the inline function into a dedicated one, you can test it easily against an http/httptest server.

Russell-IO commented:
This concept is purely awesome

MaZderMind commented:
@timoreimann Thank you for your in-depth review. It'll take me some time to get through it, though.

errm commented:
Could we add support for `X-Forwarded-For` here?

It would make sense for environments were there is some other load balancer in front of Treafik

MaZderMind commented:
I'm sure we could. How would you see that working?

Silently using the foremost Forwarded-For-Header instead of the original IP would AFAIS allow Spoofing any IP when not running behind a Proxy.

An extra Annotation would be required to turn this on and off. I'd rather like to get this feature done first and then have another issue/PR for the follow-up discussion.

timoreimann commented:
I'm with @MaZderMind: Support for the header would be great, but this PR can stand on its own without it.

timoreimann commented:
hey @MaZderMind, did you have an opportunity yet to address the comments?

I believe the feature you've built is being asked for a few times per week. Would be awesome to have it. :-)

MaZderMind commented:
Unfortunately i haven't found the time to work on. It's stil on my agenda, though.

MaZderMind commented:
@timoreimann I addressed all your change requests (despite the one on removing the empty string-slices on which I commented above) and merged the current master into this PR.

MaZderMind commented:
It seems, the build failed with "The job exceeded the maximum time limit for jobs, and has been terminated.". I'll try to retrigger.

timoreimann commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+				"8.8.8.8",
+			},
+		},
+		"IPv4 sinlge IP": {
```
Typo.

timoreimann commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+				"4.8.8.8",
+			},
+		},
+		"broken IP-adresses": {
```
typo: addresses.

timoreimann commented on middlewares/ip_witelister.go:
```diff
@@ -0,0 +1,85 @@
+import (
+	"fmt"
+	"github.com/codegangsta/negroni"
+	"github.com/containous/traefik/log"
```
Group github deps.

timoreimann commented on middlewares/ip_witelister.go:
```diff
@@ -0,0 +1,85 @@
+	for _, whitelistString := range whitelistStrings {
+		_, whitelist, err := net.ParseCIDR(whitelistString)
+		if err != nil {
+			return nil, fmt.Errorf("parsing CIDR whitelist %s: %+v", whitelist, err)
```
Error messages should just be formatted using `%v` or `%s` (no plus sign) as they are meant for humans.

timoreimann commented on server_test.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+func TestNewServerWithoutWhitelistSourceRange(t *testing.T) {
+	cases := map[string]struct {
```
Convert to struct slices here too please.

timoreimann commented on docs/toml.md:
```diff
@@ -455,6 +455,7 @@ defaultEntryPoints = ["http", "https"]
   backend = "backend1"
   passHostHeader = true
   priority = 10
+  whitelistSourceRange = ["10.42.0.0/16", "152.89.1.33/32", "afed:be44::/16"]
```
I suppose a bit of extra prose might be nice that explains things like

- what does this feature do;
- how does it behave (including error cases); and
- what providers are supported.

timoreimann commented on middlewares/ip_witelister.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (whitelister *IPWhitelister) handle(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	var match *net.IPNet
```
Unused (except for printing)?

MaZderMind commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+)
+
+func TestNewIPWhitelister(t *testing.T) {
+	cases := map[string]struct {
```
In my mind the tests are grouped by name, just as you would do it when writing them down on paper. You would not have a Headline "1" with the title as Content under it. But I'm ok following your style guide and will change this :)

timoreimann replied:
Just to add a bit more reasoning on my end: I agree with you that a map could make sense if every test had a name. However, the parameter you pass to `t.Run()` does not have to be static (as in our case) but could also be the result of some "dynamic computation". For instance, the tests for a math utility function could use the input parameters to define the description.

Looking at it from this angle, employing maps where a static name is needed and a slice of structs where it's not would lead to an inconsistent picture. I prefer tests to look boringly similar. :-)

MaZderMind commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+	}
+
+	for index, e := range cases {
+		t.Run(index, func(t *testing.T) {
```
I copied that from the docker tests, but it is there as wrong as it is here. fixed

MaZderMind commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+			if e.err != nil {
+				// expected error
+				return
+			}
```
No, line 80 checks if the returned error-value is correct. It there is an error, and the error was expected, we should not try to continue the test, because -as expected- an error has happened before.
This line ends the test when an error was to be expected.

timoreimann replied:
Ah yes, you're right of course. Thanks for the explanation.

MaZderMind commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+			//t.Parallel()
+			whitelister, err := NewIPWhitelister(e.whitelistStrings)
+
+			assert.NoError(t, err, "there should be no error")
```
I learned late about the testify assert & require packages. I changed the remaining stdlib variants to testify, because in my opinion it's much more readable

MaZderMind commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,302 @@
+
+			assert.NoError(t, err, "there should be no error")
+			if !assert.NotNil(t, whitelister, "this should not be nil") {
+				return
```
I was able to remove all but one premature test exits, and I think it does follow the correct logic here:
```
if test.err != "" {
	assert.EqualError(t, err, test.err)
	return
}
```
So IMO there is no need to introduce the extra dependency on the require module

timoreimann replied:
Hmm, I can't see why that is true. Two examples:

1. We `assert` [here](https://github.com/containous/traefik/pull/1332/files#diff-947263d47906943bdee2b9eb7cf850cdR86) that we were able to create a new IPWhitelister. Even if that assertion fails, however, we continue to iterate over the whitelister's lists. This would fail with a panic because whitelister is `nil` when `err` is non-nil. Hence, `require` seems more appropriate.
1. We do the very same assertion [here](https://github.com/containous/traefik/pull/1332/files#diff-947263d47906943bdee2b9eb7cf850cdR274) and continue to assert on the whitelister itself [one line below](https://github.com/containous/traefik/pull/1332/files#diff-947263d47906943bdee2b9eb7cf850cdR275). This is bound to fail if the previous error is already non-nil. An initial require would prevent showing a test assertion failure we don't need to see, which complicates identifying the root cause of a failed test.

(I didn't check if there are further examples.)

Besides, I don't consider the dependency on `require` to be problematic: IMHO, if it makes sense to use it, we should do so and not fall back to a rather verbose alternative.

timoreimann replied:
[This](https://github.com/containous/traefik/pull/1332/files#diff-3eb9846d5eb8762be0ee04ce73e462cbR51) and [this](https://github.com/containous/traefik/pull/1332/files#diff-3eb9846d5eb8762be0ee04ce73e462cbR55) also seem like they would benefit from `require`.

MaZderMind commented:
@timoreimann I think i changed all the things requested. The build failed with a timeout bit I'm sure retriggering it would make it succeed.

timoreimann commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,308 @@
+		desc               string
+		whitelistStrings   []string
+		expectedWhitelists []*net.IPNet
+		err                string
```
Can we rename this to `errMessage`? `err` is usually reserved for variables of type `error`.

timoreimann commented on server_test.go:
```diff
@@ -0,0 +1,61 @@
+			t.Parallel()
+			middleware, err := configureIPWhitelistMiddleware(tc.whitelistStrings)
+
+			if !reflect.DeepEqual(err, tc.err) {
```
Couldn't we use `assert.Equal()` here?

timoreimann commented on provider/string_util.go:
```diff
@@ -0,0 +1,16 @@
+
+import "strings"
+
+func splitAndTrimString(separatedString string) []string {
```
I just realize we don't have any dedicated tests for this function. Could you please add some? That way, we guarantee the function does what it should independent of any of its current users.

timoreimann commented:
@MaZderMind do you think we can manage to wrap this up before the upcoming feature freeze for the 1.3 release?

tniswong commented:
Any chance we can get the consul and consul catalog providers included in this?

MaZderMind commented:
@timoreimann I'm trying to update this PR but I really have trouble with docker_test.go which seems to have changed quite a lot upstream.

Also it seems that testify/require is not yet a dependency. Should I add it via glide, which will on the way update a lot of other dependencies (it seems glide can't just *install* something … duh).

ldez commented:
Hello, Testify is already a dependency ([glide.yaml](https://github.com/containous/traefik/blob/4ce2c8cc3409448fa95866aabf6e33381054ed9c/glide.yaml#L50)), you don't need to add him.

See [prometheus_test.go](https://github.com/containous/traefik/blob/12a0026e211eaca9d0244922421bde64acae5f32/middlewares/prometheus_test.go) for example.

timoreimann commented:
@MaZderMind I can help you deal with the rebase / docker test updates if you like. Just let me know.

MaZderMind commented:
@timoreimann I think I managed to complete the merge. Looked worse than it was.
@ldez as far as I see it's just trestify/assert, testify/require is not found in vendor (see https://github.com/containous/traefik/tree/master/vendor/github.com/stretchr/testify which does not contain https://github.com/stretchr/testify/tree/master/require)

timoreimann commented:
@MaZderMind can you try to run `./script/glide.sh get github.com/stretchr/testify/require`? It seemed to have worked for me.

If it doesn't, feel free to stick to `assert` for now. It's not ideal, but we can always fix things later.

timoreimann commented:
@tniswong this PR is already fairly large; I believe we should outsource additional providers into follow-up PRs.

Feel free to create an issue in order to track the progress (and maybe even provide a PR).

MaZderMind commented:
@timoreimann @ldez I added a Test for SplitAndTrimString and addressed all requested changes despite the switch to testify-reflect.

Running `./script/glide.sh get github.com/stretchr/testify/require` updates a ton of dependencies (https://p.rrbone.net/paste/6Whg+TJl#MSknba-uKQ/xg2HBTWbmtom7/ZPWbWEruh8pYSYeVdg) and after this, the build fails. I pushed the result including the requested uses of testify/require to https://github.com/MaZderMind/traefik/tree/ingress-ip-whitelist-require for further inspection by someone with a deeper understanding on what these dependencies actually do.

timoreimann commented:
@MaZderMind for some reason, the dependency update pulled in _older_ versions for some packages (e.g., flaeg).

I decided to pull some [cheap trick](https://github.com/containous/traefik/pull/1278#issuecomment-286696714) to update the reference to testify/require only. Things build fine on my local machine, let's see what CI says.

(I took the freedom to push my changes on top of your branch, hope you don't mind.)

timoreimann commented:
@MaZderMind one of your tests seems to be panicking. Maybe you can have a look to see if this is related to the require dependency update, the previous merge, or something different.

timoreimann commented on middlewares/ip_whitelister_test.go:
```diff
@@ -0,0 +1,307 @@
+			if test.errMessage != "" {
+				require.EqualError(t, err, test.errMessage)
+			} else {
+				for index, actual := range whitelister.whitelists {
```
Shouldn't we still have a `require.NoError` here? Otherwise, we'd only be relying on our own expectation of the error outcome through `test.errMessage`, wouldn't we?

timoreimann replied:
This one is still open.

MaZderMind replied:
Yes, sorry, I updated my last commit

timoreimann replied:
Uhm, it's still missing, isn't it?

I'd expect something like

```go
require.NoError(t, err)
```

right at the beginning of the else block starting at line 85.

MaZderMind replied:
I added it in TestIPWhitelisterHandle but missed it in TestNewIPWhitelister. That happens when you do things alongside your main work :/

fixed id and merged master one more time.

MaZderMind commented:
@timoreimann Squashed as requested. Thank you for having so much patience with me. I learned a lot throughout the review.

timoreimann commented:
@MaZderMind would you mind rebasing again? thanks!

MaZderMind commented:
@timoreimann I'm unsure how I'm supposed to merge glide.lock, as it contains a hash. naturally I'd 'glide up' glide so it can update its lockfile, but as we talked about above this would probably completely screw up the vendor versions. It's really hard to work with those vendor changes inline :/

MaZderMind commented:
@timoreimann nevermind, i manually tweaked the hash to be correct. But man, this vendor library management already caused me some headache…

timoreimann commented:
Yes, vendor management is awful. :-(

All hopes lie on golang/dep...

MaZderMind commented:
@timoreimann any news on this? I might do another rebase and merge, but it's not a task I'm very keen on, not knowing if it will be merged then.

timoreimann commented:
@MaZderMind we need another review for this PR for process reasons and due to its mere size and impact. I think it's fine to refrain from rebasing until things have been finalized.

Thanks!

emilevauge commented on middlewares/ip_witelister.go:
```diff
@@ -0,0 +1,84 @@
+func (whitelister *IPWhitelister) handle(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	remoteIP, err := ipFromRemoteAddr(r.RemoteAddr)
+	if err != nil {
+		log.Warnf("unable to parse remote-address from header: %s - rejecting", r.RemoteAddr)
```
Shouldn't you use `Debugf` everywhere (like https://github.com/containous/traefik/pull/1332/files#diff-251fc09a93a45584efee4df208e9373cR56)?

MaZderMind replied:
In my opinion this is a more dangerous problem than the others, because a single wrong syntax IP will stop the complete filter from being applied, which possibly means your private frontend is now exposed to the internet.

emilevauge replied:
Got it :+1:

emilevauge commented on middlewares/ip_witelister.go:
```diff
@@ -0,0 +1,84 @@
+
+func reject(w http.ResponseWriter) {
+	statusCode := http.StatusForbidden
+	statusText := "request from prohibited source"
```
Just asking, is it a safe practice to give a reason on this kind of error? I wonder if an empty `statusText`would't be better :thinking:

MaZderMind replied:
In my opinion, security by obscurity is never good practice. Our service should be save, iven if an attacker knows how we secured it.

emilevauge replied:
But is there a need to print this text?

MaZderMind replied:
no, not really a need. Other Webservers just repeat the HTTP Status-Code, that would probably be a string like "Forbidden". Generally I think being more verbose is better in case of debugging a problem, but If you want me to change that, I'll do so.

emilevauge replied:
@MaZderMind IMHO, we should not use HTTP error messages for debugging: we already have logs for this :) As we don't set any custom error message in Traefik for now, I would prefer to just remove it. @timoreimann WDYT?

timoreimann replied:
I'm also inclined to say there's no need to give away too much information, so either leave it empty or stick with "Forbidden" -- whatever is more consistent with what Traefik currently does.

emilevauge replied:
Then, we remove the text :) We will be able to customize the error page in the future.

emilevauge commented on server/server.go:
```diff
@@ -774,6 +782,21 @@ func (server *Server) loadConfig(configurations configs, globalConfiguration Glo
 	return serverEntryPoints, nil
 }
 
+func configureIPWhitelistMiddleware(whitelistSourceRangs []string) (negroni.Handler, error) {
```
`s/whitelistSourceRangs/whitelistSourceRanges`

MaZderMind replied:
ouch, will fix that

ldez commented:
Due to SemaphoreCI, I close and reopen the PR.

MaZderMind commented:
@emilevauge Sorry for not being responsive with this and previous changes. It's only one small side project for me. Thanks to @ldez for jumping in!
